pkg/virt-api/rest/subresource_test.go

```diff
@@ -114,9 +145,202 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+
+			request.PathParameters()["name"] = "testvmi"
+			request.PathParameters()["namespace"] = "default"
+
+			flag := false
+			vmi := v1.NewMinimalVMI("testvmi")
+			vmi.Status.Phase = v1.Running
+			vmi.ObjectMeta.SetUID(uuid.NewUUID())
+			vmi.Spec.Domain.Devices.AutoattachGraphicsDevice = &flag
+
+			server.AppendHandlers(
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("GET", "/apis/kubevirt.io/v1alpha2/namespaces/default/virtualmachineinstances/testvmi"),
+					ghttp.RespondWithJSONEncoded(http.StatusOK, vmi),
+				),
+			)
+			app.VNCRequestHandler(request, response)
+			Expect(response.Error()).To(HaveOccurred())
+			Expect(response.StatusCode()).To(Equal(http.StatusBadRequest))
+			close(done)
+		}, 5)
+
+		It("should fail with no graphics device at VNC connections", func(done Done) {
+
+			request.PathParameters()["name"] = "testvmi"
+			request.PathParameters()["namespace"] = "default"
+
+			flag := false
+			vmi := v1.NewMinimalVMI("testvmi")
+			vmi.Status.Phase = v1.Running
+			vmi.ObjectMeta.SetUID(uuid.NewUUID())
+			vmi.Spec.Domain.Devices.AutoattachGraphicsDevice = &flag
+
+			server.AppendHandlers(
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("GET", "/apis/kubevirt.io/v1alpha2/namespaces/default/virtualmachineinstances/testvmi"),
+					ghttp.RespondWithJSONEncoded(http.StatusOK, vmi),
+				),
+			)
+
+			app.VNCRequestHandler(request, response)
+			Expect(response.Error()).To(HaveOccurred())
+			Expect(response.StatusCode()).To(Equal(http.StatusBadRequest))
+			close(done)
+		}, 5)
+
+		It("Should pass client websocket io to server SPDY io", func(done Done) {
+
+			request.PathParameters()["name"] = "testvmi"
+			request.PathParameters()["namespace"] = "default"
+
+			vmi := v1.NewMinimalVMI("testvmi")
+			vmi.Status.Phase = v1.Running
+			vmi.ObjectMeta.SetUID(uuid.NewUUID())
+
+			templateService := services.NewTemplateService("whatever", "whatever", "whatever", "whatever", configCache, pvcCache)
+
+			pod, err := templateService.RenderLaunchManifest(vmi)
+			Expect(err).ToNot(HaveOccurred())
+			pod.ObjectMeta.Name = "madeup-name"
+
+			pod.Spec.NodeName = "mynode"
+			pod.Status.Phase = k8sv1.PodRunning
+
+			podList := k8sv1.PodList{}
+			podList.Items = []k8sv1.Pod{}
+			podList.Items = append(podList.Items, *pod)
+
+			newStreamChannel := make(chan httpstream.Stream)
+
+			server.AppendHandlers(
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("GET", "/apis/kubevirt.io/v1alpha2/namespaces/default/virtualmachineinstances/testvmi"),
+					ghttp.RespondWithJSONEncoded(http.StatusOK, vmi),
+				),
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("GET", "/api/v1/namespaces/default/pods"),
+					ghttp.RespondWithJSONEncoded(http.StatusOK, podList),
+				),
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("POST", "/api/v1/namespaces/default/pods/madeup-name/exec"),
+					func(w http.ResponseWriter, r *http.Request) {
+						// TODO: Just testing code we should use real SPDY stuff
```

Comment:
What do you mean by this? SPDY is deprecated and I think we shouldn't use it.

Reply:
That was something I want to ask, SPDY is deprecated but I still see it being used at kubevirt
https://github.com/kubevirt/kubevirt/blob/71454c4822b1e796d155363024a3a25be58243b9/pkg/virt-api/rest/subresource.go#L235

Reply:
k8s will switch to HTTP/2, https://github.com/kubernetes/kubernetes/issues/7452

Reply:
My bad I missed that part, I guess we will have to improve it in the future.